### PR TITLE
Area restrictions

### PR DESCRIPTION
**Pull request by lenisko**
Add area restrictions based on discord roles.

**Tested briefly.**

To get it working you need:
- filled [areas.json](https://github.com/WatWowMap/MapJS/blob/master/docs/configuration/geojson.md#scan-area) 
- correctly filled `discord.areaRestrictions` in `config.json`

For rexample:
```
"areaRestrictions": {
    "role1": {
        "roles": ["unrestrictedRoleId"],
        "areas": []
    },
    "role2": {
        "roles": ["restrictedRoleId1"],
        "areas": ["area1"]
    },
    "role3": {
        "roles": ["restrictedRoleId2"],
        "areas": ["area1", "area2"]
    },
    "role3": {
        "roles": ["restrictedRoleId3", "restrictedRoleId4"],
        "areas": ["area3"]
    }
}
```

- Any user with matching `roleId` and empty `areas` won't get restricted at all.
- User with role `restrictedRoleId1` will be able to see data only from `area1`.
- User with role `restrictedRoleId1` and `unrestrictedRoleId` won't get restricted at all.
- User with role `restrictedRoleId1` and `restrictedRoleId2` will be able to see data only from `area1, area2`.
- User with role `restrictedRoleId3` and `restrictedRoleId4` will be able to see data only from `area3`.

### CODE REVIEW

File: src/services/discord.js [hunk 1]
```diff
@@ -145,7 +149,26 @@ class DiscordClient {
                     }
                 }
             }
+            // Check once if user role is defined inside areaRestrictions
+            for (let k = 0; k < userRoles.length; k++) {
+                if (userRoles[k] in config.discord.areaRestrictions) {
+                    // Check if there's empty list for any of user roles, if so we disable restrictions
+                    if (config.discord.areaRestrictions[userRoles[k]].length === 0) overwriteAreaRestrictions = true;
```
Comment (TurtIeSocks) [inline src/services/discord.js:156]:
```suggestion
                    if (config.discord.areaRestrictions[userRoles[k]].length === 0) {
                       overwriteAreaRestrictions = true;
                    }
```

File: src/configs/config.example.json [hunk 2]
```diff
@@ -61,6 +61,11 @@
         "allowedGuilds": [],
         "blockedGuilds": [],
         "allowedUsers": [],
+        "areaRestrictions": {
```
Comment (TurtIeSocks) [inline src/configs/config.example.json:64]: I can see why you have it setup like this but I don't think it's very intuitive when compared to the rest of the config. We aren't asking admins to make changes to the keys anywhere else, just the values. I think breaking it up like the role permissions are would make more sense to the average admin that doesn't want to read the codebase to figure out what you're supposed to do. 

```suggestion
        "areaRestrictions": {
           "rule1": {
              "roles": []
              "areas": []
          },
           "rule2": {
              "roles": []
              "areas": []
          }
     }
```

File: src/services/discord.js (hunk 1 above)
Comment (TurtIeSocks) [inline src/services/discord.js:156]:
```suggestion
                    if (config.discord.areaRestrictions[userRoles[k]].length === 0) {
                       overwriteAreaRestrictions = true;
                    }
```

**Review by TurtIeSocks — CHANGES_REQUESTED** [review]
Lastly, it doesn't look like this currently works with Submission Cells and Weather. I realize those are probably a little more challenging but 🤷 overall, great work though!

File: src/services/discord.js (hunk 1 above)
Comment (TurtIeSocks) [reply src/services/discord.js:156]: okay... I don't know why it posted this one twice.

File: src/configs/config.example.json (hunk 2 above)
Comment (lenisko) [reply src/configs/config.example.json:64]: I would left current version and make documentation for this feature instead.
I've been trying to find cases in my mind, and setting more than one role to same areas list would make sense only in multi-guild cases.
With unusable keys, configuration would grow way bigger than intended unless we'll find usage for them like.. displaying access ID on WebUI or something to identify future problems. 
I'm open for discussion around that ,)

**Review by versx — APPROVED** [review]

**Review by TurtIeSocks — APPROVED** [review]
Awesome work, @lenisko !

**Pull request merged**